### DISCUSSION TITLE
Configurations regarding sound

### DISCUSSION THREAD

**Discussion by GiulianoPastorino** [dosbox-staging/dosbox-staging#3195] [Q&A] [answered]
Hi everyone,

First of all, thank you for creating and maintaining DOSBox Staging. It’s now my favorite emulator.

In the 0.79.0 release notes, you mentioned:

> **Master compressor**
An auto-levelling dynamic range compressor has been introduced to the master channel to ensure that the audio output never gets distorted, no matter how high the volume levels are. This is on by default and is virtually transparent when the output is not driven into clipping. However, purists can disable it by setting compressor to off, and then ensure correct non-clipping output levels manually.

My questions are:

1) When you say that the sound is “clipping”, you mean to say those “pops” or “clicks” that we hear when the emulator is not properly configured for a particular game?

2) If I turn the compressor off, how do I ensure correct non-clipping output levels manually? Is this done by using the mixer command? For instance, by typing `mixer sb 70`, while keeping master at 100? Or am I totally off track?

Thank you!

**Comment by johnnovak** [comment, marked as answer]
Glad to hear you're enjoying Staging @GiulianoPastorino ❤️ 

1. Digital clipping happens when the volume of the signal is digitally amplified above what your computer can handle. Digital sound reproduction has a fixed volume ceiling, and once you try to naively amplify the signal beyond that ceiling, the top of the waveforms will be "flattened", chopped off, or in other words, clipped. That results in an unpleasant distorted sound. Here is an image that illustrates the concept:

![image](https://github.com/dosbox-staging/dosbox-staging/assets/698770/f0f5378e-9417-4909-951b-74015c942a00)

![image](https://github.com/dosbox-staging/dosbox-staging/assets/698770/fef32871-4f81-44ca-889e-d31112fc1cce)

If you want to hear what clipping sounds like, you can do the following, but **DO NOT USE** headphones during experiment as you might damage your hearing!

1. Turn down the volume of your computer at the OS level to zero, or on your external amplifier or speaker
3. Set DOSBox volume to something obscenely high, say `mixer master 400` or even higher (it's game-dependent how high you need to go to make clipping really audible).
4. Start the game, then **SLOWLY** increase your OS level or external amp volume.
5. Witness the horrible noise 😎 

---

2. If you set the master volume to about 40, that will ensure that you avoid clipping in almost all circumstances. But sure, you can turn down the volume of individual mixer channels as well.

The ultimate solution would be to provide visual user feedback about clipping like on real audio mixer via clipping indicators. Note the red LEDs at the end of the LED strip—those are the clipping indicators.

![image](https://github.com/dosbox-staging/dosbox-staging/assets/698770/003f0839-9aaf-4091-a537-da61ba5d789b)

**Reply by GiulianoPastorino** [reply]
@johnnovak, thank you for your thorough explanation!

> If you want to hear what clipping sounds like, you can do the following, but DO NOT USE headphones during experiment as you might damage your hearing!

I tested it, and the sound output becomes very distorted indeed!

> If you set the master volume to about 40, that will ensure that you avoid clipping in almost all circumstances. But sure, you can turn down the volume of individual mixer channels as well.

The best choice then, for me, would be to keep `[mixer] compressor = true`, and therefore to not worry about volumes and clipping.

Here are new questions: What does generate those “pops” or “clicks” we sometimes hear in the sound, when the emulator is not properly configured for a particular game? And what are usually the best solutions for mitigating (or hopefully eliminating) them?

Thank you!

**Reply by johnnovak** [reply]
@GiulianoPastorino This section of the new Getting Started guide should help you:

https://dosbox-staging.github.io/getting-started/star-wars-dark-forces/#minimising-audio-glitches

Also the "running out of steam" part at the end of this section:

https://dosbox-staging.github.io/getting-started/star-wars-dark-forces/#setting-the-emulated-cpu-speed
